runtime/sunvmi/sunvmi.c

```diff
@@ -217,7 +217,29 @@ getCallerClassJEP176Iterator(J9VMThread * currentThread, J9StackWalkState * walk
 
 	Assert_SunVMI_mustHaveVMAccess(currentThread);
 
-	if (J9_ARE_ALL_BITS_SET(J9_ROM_METHOD_FROM_RAM_METHOD(walkState->method)->modifiers, J9AccMethodFrameIteratorSkip)) {
+#if defined(J9VM_OPT_OPENJDK_METHODHANDLE) && (JAVA_SPEC_VERSION == 11)
+/* Do not skip InjectedInvoker classes despite them having the J9AccMethodFrameIteratorSkip
+ * modifier set via the @Hidden attribute. Skipping them causes incorrect, unexpected
+ * behaviour when using OpenJDK method handles pre-hidden-class support.
+ */
+	J9_DECLARE_CONSTANT_UTF8(injectedInvokerClassname, "InjectedInvoker");
+	U_16 injectedInvokerClassnameLength = J9UTF8_LENGTH(&injectedInvokerClassname);
+	U_8 *injectedInvokerClassnameData = J9UTF8_DATA(&injectedInvokerClassname);
+
+	J9UTF8 *className = J9ROMCLASS_CLASSNAME(currentClass->romClass);
+	U_16 classNameLength = J9UTF8_LENGTH(className);
+	U_8 *classNameData = J9UTF8_DATA(className);
+
+	IDATA startIndex = classNameLength - injectedInvokerClassnameLength - ROM_ADDRESS_LENGTH - 1;
+	U_8 *start = classNameData + startIndex;
+	BOOLEAN isInjectedInvoker = ((startIndex >= 0) && (0 == memcmp(start, injectedInvokerClassnameData, injectedInvokerClassnameLength)));
```

Comment:
Is there a more performant approach to achieve this? E.g. can we cache the info associated to the `memcmp` during class file creation ... this will avoid the `memcmp` for each class here.

https://github.com/eclipse-openj9/openj9/blob/6df98f601bfd79864350b432b1266e66bfa83f00/runtime/bcutil/ClassFileWriter.hpp#L360-L374

Reply:
We can cache that the class is an InjectedInvoker during classfile writing by reserving a flag on the `extraModifiers` of the `ROMClass` associated with the `J9Class`. We can add a flag here: https://github.com/eclipse-openj9/openj9/blob/6df98f601bfd79864350b432b1266e66bfa83f00/runtime/oti/j9javaaccessflags.h#L98. `0x1000` seems unused. Then we can just check that flag in `sunvmi.c` and `NativeHelpers.cpp` from the class.

Reply:
LGTM. @TobiAjila @gacholio Do you approve this approach? Let us know if there is a better approach.

Reply:
The class name has no package. Is this intentional?

Reply:
> The class name has no package. Is this intentional?

These classes can exist in any package or any class. Only the JDK injects these classes. Our code generically targets all `InjectedInvoker` classes regardless of their package or outer/host class.

Reply:
> How do we avoid spoofed classes then? Can we check that the class has been created via Unsafe?

We use `J9AccClassAnonClass | J9AccClassHidden`.
https://github.com/eclipse-openj9/openj9/blob/6df98f601bfd79864350b432b1266e66bfa83f00/runtime/bcutil/ClassFileWriter.hpp#L353

Reply:
OK